# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: gfx/wgpu_bindings/src/client.rs
# Commit: aef2a12d7819
# Full Hash: aef2a12d7819577e36f239d5be3a248d42e34929
# Author: Dzmitry Malyshau <dmalyshau@mozilla.com>
# Date: 2022-01-05 09:33:53
# Description:
#   Bug 1746312 - Return WebGPU errors from RenderBundle creation r=jimb
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D134986
# ==============================================================================

diff -r b0072bfee83c -r aef2a12d7819 gfx/wgpu_bindings/src/client.rs
--- a/gfx/wgpu_bindings/src/client.rs	Tue Jan 04 23:00:53 2022 +0000
+++ b/gfx/wgpu_bindings/src/client.rs	Tue Jan 04 23:04:08 2022 +0000
@@ -261,6 +261,8 @@
     color_formats: *const wgt::TextureFormat,
     color_formats_length: usize,
     depth_stencil_format: Option<&'a wgt::TextureFormat>,
+    depth_read_only: bool,
+    stencil_read_only: bool,
     sample_count: u32,
 }
 
@@ -616,6 +618,7 @@
 pub extern "C" fn wgpu_device_create_render_bundle_encoder(
     device_id: id::DeviceId,
     desc: &RenderBundleEncoderDescriptor,
+    bb: &mut ByteBuf,
 ) -> *mut wgc::command::RenderBundleEncoder {
     let descriptor = wgc::command::RenderBundleEncoderDescriptor {
         label: cow_label(&desc.label),
@@ -624,15 +627,20 @@
             .depth_stencil_format
             .map(|&format| wgt::RenderBundleDepthStencil {
                 format,
-                depth_read_only: false, //TODO: add to `RenderBundleEncoderDescriptor`
-                stencil_read_only: false,
+                depth_read_only: desc.depth_read_only,
+                stencil_read_only: desc.stencil_read_only,
             }),
         sample_count: desc.sample_count,
         multiview: None,
     };
     match wgc::command::RenderBundleEncoder::new(&descriptor, device_id, None) {
         Ok(encoder) => Box::into_raw(Box::new(encoder)),
-        Err(e) => panic!("Error in Device::create_render_bundle_encoder: {}", e),
+        Err(e) => {
+            let message = format!("Error in Device::create_render_bundle_encoder: {}", e);
+            let action = DeviceAction::Error(message);
+            *bb = make_byte_buf(&action);
+            ptr::null_mut()
+        }
     }
 }
 